homeassistant/components/time_date/sensor.py

```diff
@@ -46,40 +46,68 @@ async def async_setup_platform(
     discovery_info: DiscoveryInfoType | None = None,
 ) -> None:
     """Set up the Time and Date sensor."""
-    if hass.config.time_zone is None:
-        _LOGGER.error("Timezone is not set in Home Assistant configuration")
-        return False
-
-    async_add_entities(
-        [TimeDateSensor(hass, variable) for variable in config[CONF_DISPLAY_OPTIONS]]
+    async_create_issue(
+        hass,
+        HOMEASSISTANT_DOMAIN,
+        f"deprecated_yaml_{DOMAIN}",
+        breaks_in_ha_version="2024.6.0",
+        is_fixable=False,
+        issue_domain=DOMAIN,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+        translation_placeholders={
+            "domain": DOMAIN,
+            "integration_title": "Time & Date",
+        },
     )
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Time & Date sensor."""
+    entities = []
+    for option_type in entry.options[CONF_DISPLAY_OPTIONS]:
+        entities.append(TimeDateSensor(option_type, entry.entry_id))
+    async_add_entities(entities)
 
 
 class TimeDateSensor(SensorEntity):
     """Implementation of a Time and Date sensor."""
 
-    def __init__(self, hass, option_type):
+    _attr_has_entity_name = True
+
+    def __init__(self, option_type: str, entry_id: str) -> None:
         """Initialize the sensor."""
-        self._name = OPTION_TYPES[option_type]
+        self._attr_translation_key = option_type
         self.type = option_type
-        self._state = None
-        self.hass = hass
-        self.unsub = None
+        self._state: str | None = None
```

Comment:
Outside of constructor

Reply:
What do you mean? It's in `__init__`?

Reply:
It's set to a static value, so it can be set outside of the constructor

Reply:
Right, so the opposite ;)
Tried to not touch the existing stuff but mypy complained to much 🗡️ 

btw: with this the Name becomes "Time & Date Date Time" (as example for a date & time sensor. The naming of device + sensor on this one looks really strange. Any idea?